magma/compile_guard.py

```diff
@@ -1,40 +1,79 @@
-from magma.ref import AnonRef, ArrayRef, DefnRef, InstRef, TupleRef
-from magma.t import Kind, In, Out
+from magma.ref import InstRef, get_ref_defn, get_ref_inst
+from magma.t import Kind, Type, In, Out
 from magma.type_utils import type_to_sanitized_string
-from magma.value_utils import make_selector
 
 
-def _get_top_ref(ref):
-    if isinstance(ref, TupleRef):
-        # TODO(rsetaluri): Fix port name collision.
-        return _get_top_ref(ref.tuple.name)
-    if isinstance(ref, ArrayRef):
-        # TODO(rsetaluri): Fix port name collision.
-        return _get_top_ref(ref.array.name)
-    return ref
+_logger = root_logger().getChild("compile_guard")
 
 
-@dataclasses.dataclass(frozen=True)
-class _CompileGuardState:
-    ckt: CircuitBuilder
-    ctx_mgr: contextlib.AbstractContextManager
+class _Grouper(GrouperBase):
+    def __init__(self, instances: InstanceCollection, builder: CircuitBuilder):
+        super().__init__(instances)
+        self._builder = builder
+        self._port_index = 0
+        self._clock_types = set()
+
+    def _visit_input_connection(self, _: Type, drivee: Type):
+        # NOTE(rsetaluri): Since the driver might be traced (i.e. not an
+        # immediate driver), we disregard it and grab the immediate driver
+        # (value() vs. trace()).
+        driver = drivee.value()
+        new_port = self._builder.add_port(In(type(driver)))
+        drivee.unwire(driver)
+        drivee @= new_port
+        external = getattr(self._builder, new_port.name.name)
+        external @= driver
+
+    def _visit_output_connection(self, driver: Type, drivee: Type):
+        new_port = self._builder.add_port(Out(type(driver)))
+        new_port @= driver
+        external = getattr(self._builder, new_port.name.name)
+        old_driver = drivee.value()
+        drivee.unwire(old_driver)
+        new_driver = external
+        drivee @= new_driver
+
+    def _visit_undriven_port(self, port: Type):
+        # For undriven clock types, we simply lift the port *but do not connect
+        # it* since we expect auto-wiring to do this for us (in fact, this is
+        # why the port is undriven in the first place). For non-clock types, we
+        # simply log a debug message and allow the downstream circuit pipeline
+        # to handle the undriven port.
+        if not isinstance(port, ClockTypes):
+            _logger.debug(f"found undriven port: {port}")
+            return
+        T = type(port).undirected_t
+        if T in self._clock_types:
+            return  # only add at most one port for each clock type
+        self._clock_types.add(T)
+        name = str(port.name)
+        _ = self._builder.add_port(In(T), name=name)
 
 
 class _CompileGuardBuilder(CircuitBuilder):
+    __default_defn_name_counter = 0
+
+    @staticmethod
+    def make_default_defn_name() -> str:
+        counter = _CompileGuardBuilder.__default_defn_name_counter
+        _CompileGuardBuilder.__default_defn_name_counter += 1
```

Comment:
hat tip caleb for a nice pattern to do this using [itertools.count](https://docs.python.org/3/library/itertools.html#itertools.count)

```suggestion
    __default_defn_name_counter = itertools.count()

    @staticmethod
    def make_default_defn_name() -> str:
        counter = next(_CompileGuardBuilder.__default_defn_name_counter)
```